### PR TITLE
csvdiffer replacement

### PR DESCRIPTION
**Pull request by milljm**
A working version of the new csvdiff executable.

See how far this gets through the pipline...

### CODE REVIEW

File: framework/contrib/csvdiff/csvdiff
Comment (permcody) [inline framework/contrib/csvdiff/csvdiff]: I don't think that this script should be in contrib. Contrib is for external projects that we've copied into MOOSE. If you want to open source this as a project under your profile, then it would be appropriate for it to be a contrib. As it stands it's a utility that's just part of MOOSE. I might recommend that you just put it in the scripts directory (moose/scripts).

File: framework/contrib/csvdiff/csvdiff
Comment (permcody) [inline framework/contrib/csvdiff/csvdiff]: Minor: You use set(a_line_split()) 4 times here. That's worth creating a variable so you don't have to create the set over and over.

Comment (permcody) [reply framework/contrib/csvdiff/csvdiff]: Also you have the same regex pattern in two places. You might just create the regex object once and use it twice. Helps with code maintainability too.

File: framework/contrib/csvdiff/csvdiff
Comment (permcody) [inline framework/contrib/csvdiff/csvdiff]: Convert to str then float? Why is that necessary?

File: framework/contrib/csvdiff/csvdiff
Comment (permcody) [inline framework/contrib/csvdiff/csvdiff]: Again with the double conversion. I don't think this is right or necessary.

**Review by permcody — CHANGES_REQUESTED** [review]
This is looking good. Just a few oddities in the code that should be reviewed before this goes in. That'll be good anyway since this hasn't gone through testing in a long time. We need a fresh set of testing before we can merge.

**Comment by permcody** [comment]
FYI - After we get this in, I'm going to have you clean it up a bit. All of that override stuff will just completely go away. It's a mess and won't be necessary with the new comparison file stuff.

I'm testing the differ locally with some really nasty files. I just created a custom comp file and it seems to be working well - nice!

**Review by permcody — APPROVED** [review]
This looks pretty good. I have a few more architecture things that I'd like to change, but I don't want to hold this up any longer. I'll open up separate tickets with specific tasks.

Thanks!

**Comment by bwspenc** [comment]
A couple of comments on this:

1) If we're going to use something that looks like the exodiff file format for the cmp files, we should go all the way and actually use the same file format.  The only things we need to do are to change:
`default value tolerances:` to `GLOBAL VARIABLES`, and `default time step tolerances:` to `TIME STEPS`. I just tested exodiff, and it happily runs with just these two inputs (it doesn't need the coords and all that other stuff if it's just checking globals), so if we made this change, someone would be able to use the same .cmp file to check global variables interchangeably with both exodiff and csvdiff.  This makes it so someone can generate one of these cmp files with exodiff easily if they don't happen to know where csvdiff.py is located.

2) Do we really need to deprecate the existing options for specifying individual variable tolerances in the tests file? I think it's great that we have the option for a separate .cmp file for cases when there are a lot of variables that we need to tweak. However, we often have cases where there are just one or two variables to tweak, and I prefer listing those tolerances in the tests file (like we do for global tolerances) to creating separate .cmp files.  With exodiff, we often end up with lots of .cmp files kicking around just to tweak a couple of tolerances. I actually made an issue on SEACAS a while back to support overriding tolerances for individual variables at the command line hoping that we could provide that kind of behavior for exodiff: https://github.com/gsjaardema/seacas/issues/122. Does it really hurt the maintainability of the code that much to have two ways of specifying those tolerances? We're always going to have two ways of specifying the global tolerances.

**Comment by permcody** [comment]
In response to Ben:

1. Yes! We definitely want the same format to the extent possible. BTW - Jason made a "-summary" flag for this so it _will_ generate a comparison file stub for you to work with. I agree with everything else you said.

2. You might be able to convince me on this. However, My issue is that the existing "override" options are a little screwy. They build up a set of columns seen over a set of files, they don't act on a single target. I was simply unable to get the existing options to work for my very simple case of screening out a few columns that I didn't want to check without further hacking the code. I'm open to tweaking tolerances for individual variables on the command line, but if you need to do something more complicated like look for a set of columns in different files, that kind of behavior is getting a little too complicated for command line options. Let's keep this utility small and sensible and not make it into too much of a Swiss Army knife.

**Comment by bwspenc** [comment]
@permcody I'll look into this a little more to understand how those override options behave.  If there's something about the behavior of the old tool that doesn't make sense, I'm all for changing that.

**Pull request merged**
